ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -434,26 +474,36 @@ func mutatePod(admissionReview *admissionv1.AdmissionReview) (*admissionv1.Admis
 	}
 
 	var patches []patch
-	// ray operator only sets GenerateName field - doesn't include random suffix until after admission request
-	// use mapping of {cluster name, group name, replicaIndex} -> workers to extract next TPU_WORKER_ID
-	clusterName := pod.Labels["ray.io/cluster"]
-	if clusterName == "" {
-		return nil, errors.New("Kuberay Pod missing RayCluster label")
-	}
 	containers := pod.Spec.Containers
 	if containers == nil {
 		return nil, errors.New("Container path not specified")
 	}
 	if containerRequestingTPUs(containers...) {
-		namespace := pod.Namespace
+		// ray operator only sets GenerateName field - doesn't include random suffix until after admission request
+		// use mapping of {cluster name, group name, replicaIndex} -> workers to extract next TPU_WORKER_ID
+		clusterName := pod.Labels["ray.io/cluster"]
+		if clusterName == "" {
+			return nil, errors.New("Kuberay Pod missing RayCluster label")
+		}
 		groupName := pod.Labels["ray.io/group"]
+		if groupName == "" {
+			return nil, errors.New("Kuberay Pod missing Group label")
+		}
+		namespace := pod.Namespace
 		topology := pod.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
 		if topology == "" {
 			klog.ErrorS(errors.New("TPU topology not specified"), "mutatePod", "RayCluster", namespace+"/"+clusterName, "gke-tpu-topology", topology)
 		}
 		// assign worker to the next unique ID in the pod slice and update map
 		chipsPerHost := getNumTPUChipsRequested(containers...)
 		numOfHosts, _ := getNumTPUHostsFromTopology(clusterName, groupName, namespace, topology, chipsPerHost) // ignore error here because topology may not be set yet
+
+		// query k8s client to populate sliceToWorkerIDs to then calculate the next TPU_WORKER_ID and replicaIndex
+		sliceToWorkerIDs = make(map[slice][]int)
```

Comment:
Are we locking on read too? Otherwise if you unset sliceToWorkerIDs here you could impact concurrent reads?

Reply:
Done in a707e83, I moved the `sliceToWorkerIDs`  instantiation to `updateSliceToWorkerIDs` and added a separate mutex lock for functions that read from `sliceToWorkerIDs`.

Reply:
Update: `getSliceToWorkereIDs` now creates/returns its own sliceToWorkerIDs object, which is then passed to `getReplicaIndex` and `getNextWorkerID` (and not modified anywhere else), so a mutex lock should no longer be necessary here.